src/widgets.js

```diff
@@ -229,6 +236,118 @@ define([
         return widget;
     };
 
+    var richtext = function(mug, options) {
+        var widget = normal(mug, options), editor;
+
+        function fromRichText(val) {
+            var el = $('<div>');
+            val = val.replace(/(<p>)/ig,"").replace(/<\/p>/ig, "\r\n").replace(/(<br ?\/?>)/ig,"\n");
+            el = el.html(val);
+            el.find('.atwho-inserted .label').unwrap();
+            el.find('.label-datanode').replaceWith(function() {
+                return $(this).attr('data-value');
+            });
+
+            return el.html();
+        }
+
+        function addPopovers(input) {
+            input.find('[contenteditable=false]').each(function () {
+                var $this = $(this),
+                    datavalue = $this.attr('data-value'),
+                    match = datavalue.match('output value="(.*)"'),
+                    value = match ? match[1] : $this.attr('data-value');
+                $this.popout({
+                    title: '',
+                    content: value,
+                    template: '<div contenteditable="false" class="popover"><div class="arrow"></div><div class="popover-inner"><div class="popover-content"><p></p></div></div></div>',
+                    placement: 'bottom',
+                });
+            });
+        }
+
+        function removePopovers(input) {
+            input.find('[contenteditable=false]').each(function () {
+                $(this).popout('hide');
+            });
+        }
+
+        function addCloseButton(widget, input) {
+            input.find('[contenteditable=false]').each(function () {
+                var _this = this;
+                $(this).find('.close').click(function() {
+                    $(_this).popout('hide');
+                    _this.remove();
+                    widget.handleChange();
+                    return false;
+                });
+            });
+        }
+
+        function toRichHtml(val, form, withClose) {
+            return toRichText(val, form, withClose).replace(/\r\n|\r|\n/ig, '<br />');
+        }
+
+
+        widget.input = $("<div />")
+            .attr("contenteditable", true)
+            .attr("name", widget.id)
+            .addClass('input-block-level jstree-drop');
+        if (options.singleLine) {
+            widget.input.addClass('fd-input');
+        } else {
+            widget.input.addClass('fd-textarea');
+        }
+
+        widget.input.ckeditor().promise.then(function() {
+            editor = widget.input.ckeditor().editor;
```

Comment:
This async thing feels like it's going to be a source of intermittent bugs. The widget system was not designed to handle async `getValue`/`setValue`, and I'm concerned that it's going to show up in strange places.